File: src/main/java/discretemath/bool/table/TruthTable.java
package discretemath.bool.table;

import java.util.Iterator;
import java.util.NoSuchElementException;
import java.util.Objects;

import discretemath.Matrix;
import discretemath.bool.BooleanTable;
import discretemath.bool.expression.exception.InvalidArgumentCountException;
import discretemath.combination.BitCombinations;
import discretemath.common.BitString;

public class TruthTable implements Iterable<TruthTableRow> {

	private final BooleanTable inputTable;

	private final Matrix<FunctionValue> functionValuesTable;

	private TruthTable(BooleanTable inputTable, Matrix<FunctionValue> functionValuesTable) {
		this.inputTable = inputTable;
		this.functionValuesTable = functionValuesTable;
	}

	public static TruthTable forNDegree(int n) {
		return forNDegree(n, 0);
	}

	public static TruthTable forNDegree(int n, int spaceForFunctions) {
		int rowsCount = 1 << n;
		BooleanTable table = BooleanTable.create(rowsCount, n);
		Matrix<FunctionValue> functionValueTable = Matrix.create(FunctionValue.class, rowsCount, spaceForFunctions);

		BitCombinations.nBitCombinations(n, table);

		return new TruthTable(table, functionValueTable);
	}

	public static TruthTable forNDegreeWithAllFunctions(int n) {
		int rowsCount = 1 << n;
		int functionsCount = 1 << rowsCount;
		BooleanTable table = BooleanTable.create(rowsCount, n);
		Matrix<FunctionValue> functionValueTable = Matrix.create(FunctionValue.class, rowsCount, functionsCount);

		BitCombinations.nBitCombinations(n, table);
		//BitCombinations.nBitCombinations(rowsCount, functionValueTable, 0, n, Direction.FROM_LEFT_TO_RIGHT); //TODO

		return new TruthTable(table, functionValueTable);
	}

	public void setFunctionValue(int functionIndex, int row, FunctionValue value) {
		functionValuesTable.set(row, functionIndex, value);
	}

	public void setFunctionValue(int functionIndex, FunctionValue[] values) {
		int functionValuesCount = 1 << getVariablesCount();
		InvalidArgumentCountException.checkExact(functionValuesCount, values.length);

		for (int i = 0; i < values.length; i++) {
			FunctionValue value = values[i];
			functionValuesTable.set(i, functionIndex, value);
		}

	}

	public int getVariablesCount() {
		return inputTable.columnsCount();
	}

	public int getFunctionsCount() {
		return functionValuesTable.columnsCount();
	}

	@Override
	public Iterator<TruthTableRow> iterator() {
		return new Iter();
	}

	@Override
	public boolean equals(Object o) {
		if (this == o) return true;
		if (o == null || getClass() != o.getClass()) return false;
		TruthTable that = (TruthTable) o;
		return inputTable.equals(that.inputTable) && functionValuesTable.equals(that.functionValuesTable);
	}

	@Override
	public int hashCode() {
		return Objects.hash(inputTable, functionValuesTable);
	}

	@Override
	public String toString() {
		StringBuilder builder = new StringBuilder();
		for (int i = 0; i < inputTable.rowsCount(); i++) {
			for (int j = 0; j < inputTable.columnsCount(); j++) {
				boolean b = inputTable.get(i, j);
				builder.append(b ? 1 : 0).append("\t");
			}
			builder.append("\n");
		}

		return builder.toString();
	}

	private final class Iter implements Iterator<TruthTableRow> {

		private int rowIndex = 0;

		@Override
		public boolean hasNext() {
			return rowIndex < inputTable.rowsCount();
		}

		@Override
		public TruthTableRow next() {
			if (rowIndex >= inputTable.rowsCount()) {
				throw new NoSuchElementException();
			}

			int currentRowIndex = rowIndex;
			rowIndex++;

			return new TruthTableRow() {
				@Override
				public int index() {
					return currentRowIndex;
				}

				@Override
				public int argumentsCount() {
					return getVariablesCount();
				}

				@Override
				public int functionsCount() {
					return getFunctionsCount();
				}

				@Override
				public BitString getArgumentsBitString() {
					BitString bitString = BitString.forN(inputTable.columnsCount());

					for (int j = 0; j < inputTable.columnsCount(); j++) {
						bitString.set(j, inputTable.get(currentRowIndex, j));
					}

					return bitString;
				}

				@Override
				public boolean getArgument(int index) {
					if (index > argumentsCount()) {
						throw new IndexOutOfBoundsException(index);
					}

					return inputTable.get(currentRowIndex, index);
				}

				@Override
				public FunctionValue getFunctionValue(int index) {
					if (index > functionsCount()) {
						throw new IndexOutOfBoundsException(index);
					}

					return functionValuesTable.get(currentRowIndex, index);
				}
			};
		}
	}
}

File: src/main/scala/discretemath/bool/table/TruthTable.scala
package discretemath.bool.table

import discretemath.bool.table.FunctionValue.{FALSE, TRUE}
import discretemath.combination.BitCombinations
import discretemath.common.{BitString, Direction}
import discretemath.structure.matrix.{ArrayMatrix, Matrix}

import java.util.NoSuchElementException

case class TruthTable private(private val inputTable: Matrix[Boolean], private val functionValuesTable: Matrix[FunctionValue]) extends Iterable[TruthTableRow] {

  def setFunctionValue(functionIndex: Int, row: Int, value: FunctionValue): Unit = functionValuesTable.set(row, functionIndex, value)

  def setFunctionValue(functionIndex: Int, values: Array[FunctionValue]): Unit = {
    val functionValuesCount = 1 << getVariablesCount
    assert(values.length == functionValuesCount)
    for (i <- values.indices) {
      val value = values(i)
      functionValuesTable.set(i, functionIndex, value)
    }
  }

  def getVariablesCount: Int = inputTable.columnsCount

  def getFunctionsCount: Int = functionValuesTable.columnsCount

  override def iterator: Iterator[TruthTableRow] = new Iter

  override def toString: String = {
    val builder = new StringBuilder
    for (i <- 0 until inputTable.rowsCount) {
      for (j <- 0 until inputTable.columnsCount) {
        val b = inputTable.get(i, j)
        builder.append(if (b) 1 else 0).append("\t")
      }
      builder.append("|").append("\t")
      for (j <- 0 until functionValuesTable.columnsCount) {
        val b = functionValuesTable.get(i, j)
        b match {
          case FunctionValue.TRUE => builder.append(1)
          case FunctionValue.FALSE => builder.append(0)
          case FunctionValue.DONT_CARE => builder.append('x')
        }
        builder.append("\t")
      }

      builder.append("\n")
    }
    builder.toString
  }

  final private class Iter extends Iterator[TruthTableRow] {
    private var rowIndex = 0

    override def hasNext: Boolean = rowIndex < inputTable.rowsCount

    override def next: TruthTableRow = {
      if (rowIndex >= inputTable.rowsCount) throw new NoSuchElementException
      val currentRowIndex = rowIndex
      rowIndex += 1
      new TruthTableRow() {
        override def index: Int = return currentRowIndex

        override def arguments: IndexedSeq[Boolean] = inputTable.getRow(currentRowIndex)

        def argumentsCount: Int = return getVariablesCount

        override def functionsCount: Int = return getFunctionsCount

        override def getArgumentsBitString: BitString = {
          val bitString = BitString.forN(inputTable.columnsCount)
          for (j <- 0 until inputTable.columnsCount) {
            bitString.set(j, inputTable.get(currentRowIndex, j))
          }
          bitString
        }

        override def getArgument(index: Int): Boolean = {
          if (index > argumentsCount) throw new IndexOutOfBoundsException(index)
          inputTable.get(currentRowIndex, index)
        }

        override def getFunctionValue(index: Int): FunctionValue = {
          if (index > functionsCount) throw new IndexOutOfBoundsException(index)
          functionValuesTable.get(currentRowIndex, index)
        }
      }
    }
  }
}

object TruthTable {
  def forNDegree(n: Int): TruthTable = forNDegree(n, 0)

  def forNDegree(n: Int, spaceForFunctions: Int): TruthTable = {
    val rowsCount = 1 << n
    val table: Matrix[Boolean] = ArrayMatrix.create(rowsCount, n)
    val functionValueTable: Matrix[FunctionValue] = ArrayMatrix.create(rowsCount, spaceForFunctions)
    BitCombinations.nBitCombinations(n, table)
    new TruthTable(table, functionValueTable)
  }

  def forNDegreeWithAllFunctions(n: Int): TruthTable = {
    val rowsCount = 1 << n
    val functionsCount = rowsCount << n
    val truthTable = forNDegree(n, functionsCount)
    BitCombinations.nBitCombinations(rowsCount, truthTable.functionValuesTable,
      tableFillDirection = Direction.FROM_LEFT_TO_RIGHT, transformer = value => if (value) TRUE else FALSE);

    return truthTable
  }
}